cpp/src/arrow/util/async_generator.h

```diff
@@ -1475,11 +1475,44 @@ static Result<AsyncGenerator<T>> MakeBackgroundGenerator(
   return BackgroundGenerator<T>(std::move(iterator), io_executor, max_q, q_restart);
 }
 
-/// \see MakeGeneratorIterator
-template <typename T>
+template <typename GeneratorFactory,
+          typename Generator = internal::call_traits::return_type<GeneratorFactory>,
+          typename FT = typename Generator::result_type,
+          typename T = typename FT::ValueType>
 class GeneratorIterator {
  public:
-  explicit GeneratorIterator(AsyncGenerator<T> source) : source_(std::move(source)) {}
+  explicit GeneratorIterator(GeneratorFactory factory)
+      : proxy_executor_(new internal::ProxyExecutor()) {
+    source_ = factory(proxy_executor_.get());
+  }
+
+  Result<T> Next() {
+    return internal::SerialExecutor::RunInSerialExecutor<T>(
+        [this](internal::Executor* executor) {
+          proxy_executor_->target = executor;
+          return source_();
+        });
+  }
+
+ private:
+  AsyncGenerator<T> source_;
+  std::unique_ptr<internal::ProxyExecutor> proxy_executor_;
+};
+
+template <typename GeneratorFactory,
+          typename Generator = internal::call_traits::return_type<GeneratorFactory>,
+          typename FT = typename Generator::result_type,
+          typename T = typename FT::ValueType>
+Result<Iterator<T>> MakeGeneratorIterator(GeneratorFactory factory) {
```

Comment:
It would be nice to explain the purpose of this class and whether it's reentrant or not (well, the question is moot presumably because it's blocking). This essentially 1) lets you parameterize a generator on an executor (which is guaranteed to resolve to a SerialExecutor) and then 2) synchronously blocks on the generator via the SerialExecutor, effectively running all work on the main thread?